### PR TITLE
Feature/task3

### PR DESCRIPTION
**Pull request by YahorSvidzinski**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,5 +25,11 @@ http://localhost:8080/actuator/health
 ##### http://localhost:8080/swagger-ui.html
 
 
+There is default name of your database :"postgres"
+###Second option
+#####You must be familiar with kubernetes(minukube or some cloud solutions) to load configuration
+#####I've put all configuration in kubernetes package.So don't be shy and use them as you want.
```
Comment (sergeymeshkov) [inline README.md]: Oh.... thank you so much

File: kubernetes/db-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  labels:
+    io.kompose.service: db
```
Comment (sergeymeshkov) [inline kubernetes/db-deployment.yaml]: Why those labels still here?

File: kubernetes/db-service.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    io.kompose.service: db
+  name: db
+spec:
+  type: LoadBalancer
```
Comment (sergeymeshkov) [inline kubernetes/db-service.yaml]: Really? I told 3 times about it.

File: kubernetes/web-deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        io.kompose.service: web
+    spec:
+      containers:
+      - env:
+        - name: POSTGRES_PASSWORD
```
Comment (sergeymeshkov) [inline kubernetes/web-deployment.yaml:14]: This configuration should not be in deployments file. It should be stored in secrets and configmaps. Please read documentation or https://medium.com/google-cloud/kubernetes-configmaps-and-secrets-68d061f7ab5b

File: src/main/java/com/leverx/springapp/demo/controller/UserController.java [hunk 5]
```diff
@@ -1,64 +1,67 @@
...
     @ApiResponses(value = {
-            @ApiResponse(code = 200, message = "Successfully created user"),
-            @ApiResponse(code = 400, message = "Wrong request,user name length must be between 2 and 40")
+            @ApiResponse(code = SC_CREATED, message = "Successfully created user"),
+            @ApiResponse(code = SC_BAD_REQUEST, message = "Wrong request", response = ErrorDetails.class),
+            @ApiResponse(code = SC_INTERNAL_SERVER_ERROR, message = "Server error")
     })
-    public ResponseEntity<Void> create(@Valid @RequestBody User user) {
-        userRepository.save(user);
+    @ResponseStatus(HttpStatus.CREATED)
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/controller/UserController.java]: Static import

File: src/main/java/com/leverx/springapp/demo/controller/UserController.java [hunk 6]
```diff
@@ -1,64 +1,67 @@
...
+    public ResponseEntity<Void> create(@Valid @RequestBody UserRequest userRequest) {
+        User user = userService.create(userRequest);
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/controller/UserController.java]: Task 3, 2 point - replace all to VAR

File: src/main/java/com/leverx/springapp/demo/exception/RestResponseEntityExceptionHandler.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+import static org.springframework.web.client.HttpServerErrorException.InternalServerError;
+
+@ControllerAdvice
+public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {
+
+    @ExceptionHandler(InternalServerError.class)
+    public ResponseEntity<ErrorDetails> handleError(InternalServerError ex, WebRequest webRequest) {
+        String message = "Server error";
+        ErrorDetails errorDetails = new ErrorDetails(message, HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/exception/RestResponseEntityExceptionHandler.java]: STATIC IMPORT

File: src/main/java/com/leverx/springapp/demo/exception/RestResponseEntityExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        return new ResponseEntity(errorDetails, errorDetails.getHttpStatus());
+    }
+
+    @ExceptionHandler({ ConstraintViolationException.class })
+    public ResponseEntity<ErrorDetails> handleConstraintViolation(
+            ConstraintViolationException ex, WebRequest request) {
+        String message = "Wrong user names,length must be between 2 and 40";
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/exception/RestResponseEntityExceptionHandler.java]: No. Again - This message should **not be harcoded**. You should get information from exception. If we will add additional field 'body type' with different constraints, this handler will always return error "Wrong user names,length must be between 2 and 40", even if name if correct.

File: src/main/java/com/leverx/springapp/demo/exception/RestResponseEntityExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    @ExceptionHandler(InternalServerError.class)
+    public ResponseEntity<ErrorDetails> handleError(InternalServerError ex, WebRequest webRequest) {
+        String message = "Server error";
+        ErrorDetails errorDetails = new ErrorDetails(message, HttpStatus.INTERNAL_SERVER_ERROR);
+        return new ResponseEntity(errorDetails, errorDetails.getHttpStatus());
+    }
+
+    @ExceptionHandler({ ConstraintViolationException.class })
+    public ResponseEntity<ErrorDetails> handleConstraintViolation(
+            ConstraintViolationException ex, WebRequest request) {
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/exception/RestResponseEntityExceptionHandler.java]: Where you use it 'WebRequest request'?

File: src/main/java/com/leverx/springapp/demo/exception/RestResponseEntityExceptionHandler.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        String message = "Wrong user names,length must be between 2 and 40";
+        ErrorDetails errorDetails =
+                new ErrorDetails(message, HttpStatus.BAD_REQUEST);
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/exception/RestResponseEntityExceptionHandler.java]: Egor, error code != HTTP error code. This is separate consents. HTTP error code says common info - for example that request body is incorrect. But we DON'T know what exaclty incorrect. Error code in this entity is business code. It should say, for example, that code 0001 - First name is incorrect, 0002 - Second name is incorrect, 0003 - body type is incorrect and etc.

File: src/main/java/com/leverx/springapp/demo/repository/CommonUserRepository.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package com.leverx.springapp.demo.repository;
+
+import com.leverx.springapp.demo.model.User;
+
+import java.util.Collection;
+import java.util.Optional;
+
+public interface CommonUserRepository {
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/repository/CommonUserRepository.java]: Lets name it just 'UserRepository'

File: src/main/java/com/leverx/springapp/demo/repository/UserJdbcRepositoryImpl.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+
+@Repository
+@RequiredArgsConstructor
+public class UserJdbcRepositoryImpl implements CommonUserRepository {
+
+    public static final String SELECT_ID_FROM_USERS_ORDER_BY_ID_DESC_LIMIT_1 = "SELECT id FROM users ORDER BY id DESC LIMIT 1";
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/repository/UserJdbcRepositoryImpl.java]: Why some SQL statements in constants, and some of them not??

File: src/main/java/com/leverx/springapp/demo/repository/UserJdbcRepositoryImpl.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    public static final String INSERT_INTO_USERS_FIRST_NAME_SECOND_NAME_VALUES = "INSERT INTO users (first_name, second_name) values (?,?)";
+    public static final String SELECT_ID_FIRST_NAME_SECOND_NAME_FROM_USERS_WHERE_ID = "SELECT id, first_name, second_name FROM users WHERE id = ?";
+    private final JdbcTemplate jdbcTemplate;
+
+    @Override
+    public User save(User user) {
+        String sql = INSERT_INTO_USERS_FIRST_NAME_SECOND_NAME_VALUES;
+        jdbcTemplate.update(sql, user.getFirstName(), user.getSecondName());
+        sql = SELECT_ID_FROM_USERS_ORDER_BY_ID_DESC_LIMIT_1;
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/repository/UserJdbcRepositoryImpl.java]: It's not correct approach
https://docs.oracle.com/javase/6/docs/api/java/sql/Statement.html#RETURN_GENERATED_KEYS

File: src/main/java/com/leverx/springapp/demo/repository/UserJdbcRepositoryImpl.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    public Optional<User> findById(Long id) {
+        String sql = SELECT_ID_FIRST_NAME_SECOND_NAME_FROM_USERS_WHERE_ID;
+        RowMapper<User> rowMapper = new BeanPropertyRowMapper<User>(User.class);
+        User user = jdbcTemplate.queryForObject(sql, rowMapper, id);
+        return Optional.ofNullable(user);
+    }
+
+    @Override
+    public List<User> findAll() {
+        String sql = "SELECT id, first_name, second_name FROM users";
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/repository/UserJdbcRepositoryImpl.java]: Why you used different formatting for different SQL statements?

File: src/main/java/com/leverx/springapp/demo/service/UserService.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+import com.leverx.springapp.demo.model.User;
+import com.leverx.springapp.demo.model.UserRequest;
+
+import javax.validation.Valid;
+import java.util.Collection;
+import java.util.Optional;
+
+public interface UserService {
+
+    User create(@Valid UserRequest userRequest);
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/service/UserService.java:14]: Allow Null Values? Please add @NotNull

File: src/main/java/com/leverx/springapp/demo/service/UserServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Collection;
+import java.util.Optional;
+
+import static org.springframework.transaction.annotation.Isolation.REPEATABLE_READ;
+
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+
+    private final CommonUserRepository userRepository;
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/service/UserServiceImpl.java]: Please remove empty spaces

File: src/main/java/com/leverx/springapp/demo/service/UserServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+
+    @Override
+    @Transactional
+    public User create(@Valid UserRequest userRequest) {
+        User user = new User();
+        BeanUtils.copyProperties(userRequest, user);
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/service/UserServiceImpl.java:26]: Ok, I don't like this approach by lt's leave it or now.

File: src/main/java/com/leverx/springapp/demo/controller/UserController.java [hunk 18]
```diff
@@ -1,64 +1,67 @@
...
-            @ApiResponse(code = 200, message = "Successfully created user"),
-            @ApiResponse(code = 400, message = "Wrong request,user name length must be between 2 and 40")
+            @ApiResponse(code = SC_CREATED, message = "Successfully created user"),
+            @ApiResponse(code = SC_BAD_REQUEST, message = "Wrong request", response = ErrorDetails.class),
+            @ApiResponse(code = SC_INTERNAL_SERVER_ERROR, message = "Server error")
     })
-    public ResponseEntity<Void> create(@Valid @RequestBody User user) {
-        userRepository.save(user);
+    @ResponseStatus(HttpStatus.CREATED)
+    public ResponseEntity<Void> create(@Valid @RequestBody UserRequest userRequest) {
```
Comment (sergeymeshkov) [inline src/main/java/com/leverx/springapp/demo/controller/UserController.java:37]: Are you sure that @Valid annotation will work here? I don't. Please test your code.

File: README.md [hunk 19]
```diff
@@ -1,20 +1,22 @@
-# TASK2
+# TASK3
```
Comment (sergeymeshkov) [inline README.md:1]: Still 15 changes in 1 commit

**Review by sergeymeshkov — CHANGES_REQUESTED** [review]

File: kubernetes/db-service.yaml (hunk 3 above)
Comment (YahorSvidzinski) [reply kubernetes/db-service.yaml]: Outdated

**Pull request closed without merging**